framework/src/org/checkerframework/common/signatureinference/SignatureInferenceScenes.java

```diff
@@ -0,0 +1,726 @@
+    }
+
+    /**
+     * Updates the set of annotations in a location of a Scene.
+     *   <ul>
+     *     <li>If there was no previous annotation for that location, then the
+     *      updated set will be the annotations in newATM.</li>
+     *     <li>If there was a previous annotation, the updated set will be the
+     *      LUB between the previous annotation and newATM.</li>
+     *   </ul>
+     * <p>
+     * @param type ATypeElement of the Scene which will be modified.
+     * @param atf the annotated type factory of a given type system, whose
+     * type hierarchy will be used.
+     * @param jaifPath used to identify a Scene.
+     * @param newATM the type containing annotations to be added (or "lubbed")
+     * to the Scene.
+     * @param defLoc the location where the annotation will be added.
+     */
+    private static void updateAnnotationSetInScene(ATypeElement type,
+            AnnotatedTypeFactory atf, String jaifPath,
+            AnnotatedTypeMirror newATM, AnnotatedTypeMirror oldATM,
+            DefaultLocation defLoc) {
+        AnnotatedTypeMirror curAnnosATM = typeElementToATM(
+                type, atf, newATM.getUnderlyingType());
+        if (curAnnosATM.getAnnotations().size() == newATM.getAnnotations().size()
+                && !newATM.getAnnotations().isEmpty()) {
+            newATM = AnnotatedTypes.leastUpperBound(
+                    atf.getProcessingEnv(), atf, newATM, curAnnosATM);
+        }
+        updateTypeElementFromATM(newATM, oldATM, atf, type, 1, defLoc);
+        modifiedScenes.add(jaifPath);
+    }
+
+    /**
+     * Returns the AClass in an AScene, given a className and a jaifPath.
+     */
+    private static AClass getAClass(String className, String jaifPath) {
+        // Possibly reads .jaif file to obtain a Scene.
+        AScene scene = getScene(jaifPath);
+        return scene.classes.vivify(className);
+    }
+
+    /**
+     * Returns true if am should not be inserted in source code. This happens
+     * when am cannot be inserted in source code or is the default for the
+     * location passed as argument.
+     * <p>
+     * For example, {@link org.checkerframework.common.value.qual.BottomVal}.
+     * Returns false otherwise.
+     * <p>
+     * Invisible qualifiers, which are annotations that contain the
+     * {@link org.checkerframework.framework.qual.InvisibleQualifier}
+     * meta-annotation, also return true.
+     */
+    private static boolean shouldIgnore(AnnotationMirror am,
+            DefaultLocation location) {
+        boolean shouldIgnore = false;
+        // Checks if am is an implementation detail
+        Target target = am.getAnnotationType().asElement().
+                getAnnotation(Target.class);
+        shouldIgnore = shouldIgnore || (target != null && target.value().length == 0);
+        shouldIgnore = shouldIgnore || am.getAnnotationType().asElement().
+                getAnnotation(InvisibleQualifier.class) != null;
+
```

Comment:
As an optimization, you could skip the expensive operations that follow if shouldIgnore is true at this point.